File: rust/src/client_pool.rs
use std::num::NonZeroUsize;
use std::sync::{Arc, Mutex};
use async_trait::async_trait;
use lru::LruCache;
use tokio::sync::Mutex as TokioMutex;

const CLIENT_CACHE_SIZE: usize = 20;

#[async_trait]
pub trait Connect: Clone + Send + Sync + 'static {
    type Error;
    async fn connect(endpoint: String) -> Result<Self, Self::Error>;
}

#[derive(Debug, Clone)]
pub struct ClientPool<T> {
    client_cache: Arc<TokioMutex<LruCache<String, T>>>,
}

impl<T> ClientPool<T>
where
    T: Connect,
{
    pub fn new() -> Self {
        Self {
            client_cache: Arc::new(TokioMutex::new(LruCache::new(
                NonZeroUsize::new(CLIENT_CACHE_SIZE).unwrap(),
            ))),
        }
    }

    pub async fn connect(&self, address: &str) -> Result<T, T::Error> {
        let client = T::connect(address.to_string()).await?;
        self.client_cache
            .lock()
            .await
            .put(address.to_string(), client.clone());
        Ok(client)
    }

    pub async fn get_client(&self, address: &str) -> Result<T, T::Error> {
        if let Some(client) = self.client_cache.lock().await.get(address) {
            return Ok(client.clone());
        }
        self.connect(address).await
    }
}

File: go/client_pool.go
package futures

import (
    "sync"
    lru "github.com/hashicorp/golang-lru"
)

const ClientCacheSize = 20

type Connectable interface {
    Connect(address string) (interface{}, error)
}

type ClientPool struct {
    clientCache *lru.Cache
    mutex       sync.Mutex
}

func NewClientPool() (*ClientPool, error) {
    cache, err := lru.New(ClientCacheSize)
    if err != nil {
        return nil, err
    }
    return &ClientPool{
        clientCache: cache,
    }, nil
}

func (p *ClientPool) Connect(c Connectable, address string) (interface{}, error) {
    p.mutex.Lock()
    defer p.mutex.Unlock()

    if client, ok := p.clientCache.Get(address); ok {
        return client, nil
    }

    client, err := c.Connect(address)
    if err != nil {
        return nil, err
    }

    p.clientCache.Add(address, client)
    return client, nil
}

func (p *ClientPool) GetClient(c Connectable, address string) (interface{}, error) {
    p.mutex.Lock()
    defer p.mutex.Unlock()

    if client, ok := p.clientCache.Get(address); ok {
        return client, nil
    }

    return p.Connect(c, address)
}